spec.bs

```diff
@@ -572,27 +570,51 @@ An <dfn>app history API navigation</dfn> is a [=struct=] with the following [=st
+  1. Let |promise| be [=a new promise=] created in |appHistory|'s [=relevant Realm=].
+
+  1. Let |traversal| be an [=app history API navigation=] whose [=app history API navigation/key=] is |key|, [=app history API navigation/info=] is |info|, [=app history API navigation/serialized state=] is null, and [=app history API navigation/returned promise=] is |promise|.
 
-  1. Set |appHistory|'s [=AppHistory/ongoing non-traverse navigation=] to |appHistory|'s [=AppHistory/upcoming non-traverse navigation=].
+  1. Set |appHistory|'s [=AppHistory/upcoming traverse navigations=][|key|]  to |traversal|.
 
-  1. Set |appHistory|'s [=AppHistory/upcoming non-traverse navigation=] to null.
+  1. Return |traversal|.
 </div>
 
 <div algorithm>
-  To <dfn for="AppHistory">set an ongoing traverse navigation</dfn> given an {{AppHistory}} |appHistory|, a string |key|, and a JavaScript value |info|:
+  To <dfn for="AppHistory">promote the upcoming navigation to ongoing</dfn> given an {{AppHistory}} |appHistory| and a string-or-null |destinationKey|:
 
-  1. Let |promise| be [=a new promise=] created in |appHistory|'s [=relevant Realm=].
+  1. Assert: |appHistory|'s [=AppHistory/ongoing navigation=] is null.
 
-  1. Let |cleanupStep| be an algorithm step which [=map/removes=] |appHistory|'s [=AppHistory/ongoing traverse navigations=][|key|].
+  1. If |destinationKey| is not null, then:
 
-  1. Let |ongoingNavigation| be an [=app history API navigation=] whose [=app history API navigation/info=] is |info|, [=app history API navigation/serialized state=] is null, [=app history API navigation/returned promise=] is |promise|, and [=app history API navigation/cleanup step=] is |cleanupStep|.
+    1. Assert: |appHistory|'s [=AppHistory/upcoming non-traverse navigation=] is null.
 
-  1. Set |appHistory|'s [=AppHistory/ongoing traverse navigations=][|key|]  to |ongoingNavigation|.
+    1. If |appHistory|'s [=AppHistory/upcoming traverse navigations=][|destinationKey|] [=map/exists=], then:
 
-  1. Return |ongoingNavigation|.
+      1. Set |appHistory|'s [=AppHistory/ongoing navigation=] to |appHistory|'s [=AppHistory/upcoming traverse navigations=][|destinationKey|].
+
+      1. [=map/Remove=] |appHistory|'s [=AppHistory/upcoming traverse navigations=][|destinationKey|].
+
+  1. Otherwise,
+
+    1. Set |appHistory|'s [=AppHistory/ongoing navigation=] to |appHistory|'s [=AppHistory/upcoming non-traverse navigation=].
+
+    1. Set |appHistory|'s [=AppHistory/upcoming non-traverse navigation=] to null.
+</div>
+
+<div algorithm>
+  To <dfn for="AppHistory">clean up an API navigation</dfn> given an {{AppHistory}} |appHistory| and an [=app history API navigation=] |navigation|:
+
+  1. If |appHistory|'s [=AppHistory/ongoing navigation=] is |navigation|, then set |appHistory|'s [=AppHistory/ongoing navigation=] to null.
+
+  1. Otherwise,
+
+    1. Assert: |navigation|'s [=app history API navigation/key=] is not null.
```

Comment:
This assert is stronger than the implementation. Let me know if it's incorrect.

Reply:
I think you're right.